src/libcore/result.rs

```diff
@@ -684,11 +684,16 @@ impl<T, E: fmt::Debug> Result<T, E> {
     pub fn unwrap(self) -> T {
         match self {
             Ok(t) => t,
-            Err(e) =>
-                panic!("called `Result::unwrap()` on an `Err` value: {:?}", e)
+            Err(e) => Self::unwrap_failed(e),
         }
     }
 
+    #[inline(never)]
+    #[cold]
+    fn unwrap_failed(error: E) -> ! {
+        panic!("called `Result::unwrap()` on an `Err` value: {:?}", error)
+    }
```

Comment:
Yes. I've only compared taking `E` vs `&E`. Taking `E` has the benefit that the deallocation code (for String or std::io::Error) moves into the cold path as well..